### ISSUE TITLE
ENH: Support obects with __array__ method

### ISSUE THREAD

**Issue by snowman2** [pyproj4/pyproj#573]
https://docs.scipy.org/doc/numpy/reference/arrays.interface.html

Had the thought, putting here to look into.

**Comment by mraspaud** [comment]
Yes, it would be great to have support for `xarray.DataArray` or `dask.array.Array` objects!

**Comment by snowman2** [comment]
I did some looking into `xarray.DataArray` and is seems it does not have `__array_interface__`. However, it does have [__array__](https://github.com/pydata/xarray/blob/6ce07249ca7eabc181b2b88a5723e66fc06036d4/xarray/core/common.py#L131-L132). @mraspaud did you have any thoughts on what behavior you would like to see there?

**Comment by mraspaud** [comment]
Oh, sorry, I hadn't checked that xarray had this interface. Anyway, our use case is that we have `xarray.DataArray` structures containing `dask.array.Array` objects. Passing any of these to a `Proj` object just fails at the moment:
```python
 In [1]: import xarray as xr                                                     

In [2]: from pyproj import Proj                                                 

In [6]: prj(xr.DataArray([1, 2, 3]), xr.DataArray([1, 2, 3]))                   

TypeError: input must be an array, list, tuple or scalar

In [7]: import dask.array as da                                                 

In [8]: prj(da.from_array([1, 2, 3]), da.from_array([1, 2, 3]))                 

TypeError: input must be an array, list, tuple or scalar
```
It would be nice if pyproj wouldn't just choke on these. Just calling eg `np.asarray` should suffice. Now I realise pyproj uses the underlying proj library and thus converts array to C structures at one point, so it would loose the nice features of xarray or dask along the way, but it would work.

**Comment by snowman2** [comment]
Okay, so if I interpreted that correctly, you would be okay with getting a numpy array returned?

Next step is to check if dask has the same `__array__` method.

**Comment by mraspaud** [comment]
I'd be ok with that. Maybe @djhoese has an opinion on this ?

**Comment by djhoese** [comment]
Part of me wishes pyproj could magically know how to handle dask arrays in a dask-friendly way (map_blocks wrapper), but I think converting to a numpy array and returning a numpy array matches xarray behavior. For example, when there are methods in xarray that dask doesn't support I believe xarray just converts to numpy and continues on processing.

**Comment by snowman2** [comment]
Looks like the dask array also has [\_\_array__](https://github.com/dask/dask/blob/9968a49bb4e9ae03d29110bbc5a41a4179636ab8/dask/array/core.py#L1336), so I think the code could be updated to just call that function and it would work for both.

**Comment by snowman2** [comment]
Implemented in #625

**Comment by snowman2** [comment]
Looks like this will also work with [pandas.Series.\_\_array__](https://github.com/pandas-dev/pandas/blob/3ed7dff48bb4e8c7c0129283ff51eccea3a0f861/pandas/core/series.py#L752)
